cmd/controller/main.go

```diff
@@ -68,22 +88,80 @@ func main() {
 	flag.Parse()
 	logf.SetLogger(logf.ZapLogger(false))
 
-	// Read the logging config and setup a logger.
-	cm := getLoggingConfigOrDie()
-
-	config, err := logging.NewConfigFromMap(cm, logconfig.Controller)
-	if err != nil {
-		log.Fatalf("Error parsing logging configuration: %v", err)
-	}
-	logger, atomicLevel := logging.NewLoggerFromConfig(config, logconfig.Controller)
+	logger, atomicLevel := setupLogger()
 	defer logger.Sync()
 	logger = logger.With(zap.String(logkey.ControllerType, logconfig.Controller))
 
-	logger.Info("Starting the controller")
-
 	// set up signals so we handle the first shutdown signal gracefully
 	stopCh := signals.SetupSignalHandler()
 
+	go startPkgController(stopCh, logger, atomicLevel)
+	go startControllerRuntime(stopCh, logger, atomicLevel)
+}
+
+func startPkgController(stopCh <-chan struct{}, logger *zap.SugaredLogger, atomicLevel zap.AtomicLevel) {
+	logger = logger.With(zap.String("controller/impl", "pkg"))
+	logger.Info("Starting the controller")
+
+	cfg, err := clientcmd.BuildConfigFromFlags(*masterURL, *kubeconfig)
+	if err != nil {
+		logger.Fatalw("Error building kubeconfig", zap.Error(err))
+	}
+
+	const numControllers = 1
```

Comment:
why not define this with the other constants around L72?

Reply:
I am copying patterns from serving, I think because this code knows how many reconcilers it is adding right here.

Reply:
Why can't we determine the qps values based on the size of the controller array?

Reply:
I was going to suggest that as well, but these seem to be necessary before the controller array is defined. Then I was going to suggest that we should document below in the controller array creation, but then I noticed that the code will panic if you add a controller and forget to modify controller const here. So I'm fine with this.

Reply:
This is a copy from serving. We can improve on it later. I would suggest not changing it until we do more integrations.